### DISCUSSION TITLE
Connecting to Google Cloud SQL

### DISCUSSION THREAD

**Discussion by ryardley** [vercel/vercel#4595] [Help] [answered]
How do I connect a Vercel serverless function say on a Next.JS app to Google Cloud Platform Cloud SQL Postgres instance? 
It seems I need to either:
1. Upload a secret file like a certificate
2. Run the google cloud proxy

I can only find the ability to use environment vars to send secrets to instances so that doesn't seem like the right way.

I noticed there is an integration that looks written by vercel here: https://github.com/vercel/integrations/tree/master/gcloud

This seems to provide IP / Username / Password for my SQL database.

Does this run the proxy in production? Scouring through the code I couldn't workout easily whether or not that is the case although it didn't look like it.

Pushing to production using the details provided by the integration did not appear to work either.

I assume I cannot whitelabel vercel serverless function IPs.

At this point this seems to me like it is not supported although it appears that it might be because of the integration.

**Comment by mcsdevv** [comment, marked as answer]
Hey @ryardleyl, the easiest way to do this would be option 1, whilst adding an allowlist of 0.0.0.0 to GCP for the Serverless Functions IP addresses. Vercel deployments have dynamic IP addresses and so that it is required to allow all for the connection to work correctly.

**Reply by alex-hladun** [reply]
Just a note for anyone stumbling on this, you actually need to add 0.0.0.0/0 to the Authorized networks

**Reply by villesau** [reply]
Hi,

How does one exactly upload secret files in Vercel? All I can see is to setup env variables. However, Prisma requires ssl certs to be files: https://www.prisma.io/docs/concepts/database-connectors/postgresql#arguments

Also even all the hacks with env variables runs out of space.

**Reply by brian-dlee** [reply]
There are some answers scattered throughout the web on how to get your SSL certificates securely into your Vercel deployment for Prisma. There are a few issues to overcome with encrypting the certificates, getting them into your environment variables, making sure the encrypted file you write gets included in your build, and writing them to tmp files on startup to appease Prisma. I'd seen enough code snippets and found enough resources to piece together a more turn-key solution that maybe the community could benefit from.

https://github.com/brian-dlee/smuggler

**Reply by amyegan** [reply]
Thanks for sharing this solution @brian-dlee!

**Reply by Beetroot11** [reply]
2 years later - still an issue - thanks again @alex-hladun

**Comment by paulogdm** [comment]
Another reference: https://vercel.com/knowledge/how-to-allowlist-deployment-ip-address

**Comment by diegoxro** [comment]
I read the article shared above , are the IPs really too dynamic to come up with a rule that would cover the IPs vercel uses ?

**Reply by paulogdm** [reply]
It is simply not safe to assume a range at the moment. Not only that, but we recommend that you deploy APIs to serverful environments so you can have more control over the IPs. Then on your Vercel app, you can query the API with an authorization token or something similar.

**Reply by diegoxro** [reply]
"at the moment" , then someday we could "assume" them ? I dont want to control the IPs , basically what you are saying is "dont use vercel" to deploy your APIs

**Reply by paulogdm** [reply]
That is correct. Full-blown APIs should not be deployed in Vercel. Serverless functions should be "helpers" and "add-ons" to your front-end, which is our current focus today.

**Reply by aecorredor** [reply]
@paulogdm hey man, that's a pretty big statement. I know that with the new rebranding FE is the focus, but a lot of us that were in zeit now from before and had full blown APIs would be really surprised from only reading this deep within a discussion. I think it would be fair to clearly advertise exactly that somewhere in the docs so people know. Specifically `Serverless functions should be "helpers" and "add-ons" to your front-end, which is our current focus today.`.

**Reply by DavidChouinard** [reply]
@paulogdm is this still the case that "full-blown APIs should not be deployed in Vercel"?

**Reply by leerob** [reply]
I'd say there's a bit more nuance. One thing to call out, in particular, is [Edge Functions](https://vercel.com/features/edge-functions).

Edge Functions help solve some of the tradeoffs of building an API with Serverless Functions, like eliminating long cold starts, and being generally more cost-effective at scale due to the smaller runtime. This isn't without its own tradeoffs, as the Edge ecosystem is not as feature-rich as the much larger Node.js ecosystem, but it can be a viable option for certain types of standalone APIs.

[Here's an example of an API built using Edge Functions from our customer Planetscale.](https://planetscale.com/blog/we-now-display-planetscale-system-status-directly-in-your-dashboard)

Overall, the general sentiment of a large-scale API (that has, let's say, 50+ people working on it, with all of the logging, monitoring, rate-limiting, and other goodies to consider) not being a good fit for Vercel is still mostly accurate. While [monitoring](https://vercel.com/changelog/monitoring-is-in-public-beta-for-enterprise-teams) and [logging](https://vercel.com/changelog/improved-logs-available-as-public-beta-for-enterprise-teams) on Vercel is improving drastically, and even [solutions for rate limiting](https://vercel.com/templates/next.js/api-rate-limit-and-tokens), it's ultimately more of a function for how the team wants to operate.

**Reply by DavidChouinard** [reply]
Awesome very helpful thank you @leerob 

We run a large application (1000+ endpoints) on Vercel serverless functions but it's been good so far. Perhaps the biggest benefit has been being able to deploy client-side and server-side changes all at once and have deploy previews that include both the client and server-side changes for a feature.

**Reply by angusryer** [reply]
@DavidChouinard Is this still true today? 1000+ endpoints running as serverless functions is a great metric for people to know when exploring if their application can be deployed using this arch.

**Reply by DavidChouinard** [reply]
Yes, it is. Can vouch it's solid.

**Comment by Aleksion** [comment]
Did anyone ever get the google cloud integration to work? I've tried quite a few combinations of it without success:

I'm using prisma, but here's the connection string I've setup:

postgresql://SQL_PROD_USERNAME:SQL_PROD_PASSWORD@SQL_PROD_IP:3306/marketing-backend?host=/cloudsql/[INSTANCE_CONNECTION_STRING]

I originally had a google app engine web server serving requests, but nextjs with getstaticprops brings it to its knees. So I wanted to connect directly to the db for the builds instead

**Reply by maxekman** [reply]
Were you able to connect with this string from a Vercel deploy (non-GCP cloud run)? I have managed to get the secrets into the Vercel project, but I have no idea how to create an env var that expands the secrets into a Prisma DATABASE_URL connection string.

**Reply by joshuacrowley** [reply]
@maxekman did you figure it out? Just trying that now.

**Reply by maxekman** [reply]
I sadly ended up using a public instance without allowed networks (0.0.0.0). For testing/dev I didn't even bother with SSL so I have no input on that either right now.

**Reply by maxekman** [reply]
Googles SQL solution is really good when staying inside their networking, but I suspect Heroku/other solutions could fit this external use case better.

**Reply by cameronembling** [reply]
Take a look at digital ocean managed databases, they seem to be easier to reason about for serverless first apps
